torch/testing/_internal/distributed/rpc/dist_autograd_test.py

```diff
@@ -1169,6 +1169,39 @@ def test_backward_autograd_engine_error(self):
                 # Run backwards, and validate we receive an error.
                 dist_autograd.backward(context_id, [val.sum()])
 
+    @dist_init
+    def test_backward_intermediate_autograd_engine_error(self):
+        with dist_autograd.context() as context_id:
+            t1 = torch.rand((3, 3), requires_grad=True)
+            t2 = torch.rand((3, 3), requires_grad=True)
+            # Perform some ops before error simulation.
+            tmp = (t1 + t2) * (t1 + t2)
+            t3 = SimulateBackwardError.apply(tmp)
+
+            # Run multiple round trips across different nodes and verify the
+            # original node receives an error thrown at some intermediate node
+            # in the chain.
+            val = rpc.rpc_sync(
+                "worker{}".format(self._next_rank()), torch.add, args=(t2, t1)
+            )
+            val = rpc.rpc_sync(
+                "worker{}".format(self._next_rank()), torch.mul, args=(val, t3)
+            )
+            val = rpc.rpc_sync(
+                "worker{}".format(self._next_rank()), torch.matmul, args=(val, t2)
+            )
+            val = rpc.rpc_sync(
+                "worker{}".format(self._next_rank()), torch.div, args=(val, t2)
+            )
+
+            with self.assertRaisesRegex(
+                RuntimeError, "(Simulate error on backward pass)|(RPC ran for more than)"
+            ):
+                # Run backwards, and validate we receive an error.
+                dist_autograd.backward(context_id, [val.sum()])
+
```

Comment:
Can we add the number of `DIST_AUTOGRAD_FAILURE_REQ` messages received per context id to debug info and then verify here that each node gets one of these messages for all the other context ids? Can add this as a separate PR.